tools/local-sensor/main.go

```diff
@@ -0,0 +1,158 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"os/signal"
+	"path"
+	"strconv"
+	"syscall"
+	"time"
+
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/utils"
+	centralDebug "github.com/stackrox/rox/sensor/debugger/central"
+	"github.com/stackrox/rox/sensor/debugger/k8s"
+	"github.com/stackrox/rox/sensor/debugger/message"
+	"github.com/stackrox/rox/sensor/kubernetes/sensor"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+// local-sensor is an application that allows you to run sensor in your host machine, while mocking a
+// gRPC connection to central. This was introduced for testing and debugging purposes. At its current form,
+// it does not connect to a real central, but instead it dumps all gRPC messages that would be sent to central in a file.
+
+func createConnectionAndStartServer(fakeCentral *centralDebug.FakeService) (*grpc.ClientConn, *centralDebug.FakeService, func()) {
+	buffer := 1024 * 1024
+	listener := bufconn.Listen(buffer)
+
+	server := grpc.NewServer()
+	central.RegisterSensorServiceServer(server, fakeCentral)
+
+	go func() {
+		utils.IgnoreError(func() error {
+			return server.Serve(listener)
+		})
+	}()
+
+	conn, err := grpc.DialContext(context.Background(), "", grpc.WithContextDialer(func(ctx context.Context, s string) (net.Conn, error) {
+		return listener.Dial()
+	}), grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	if err != nil {
+		panic(err)
+	}
+
+	closeF := func() {
+		utils.IgnoreError(listener.Close)
+		server.Stop()
+	}
+
+	return conn, fakeCentral, closeF
+}
+
+func mustGetCommandLineArgs() (time.Duration, string) {
+	if len(os.Args) != 3 {
+		fmt.Println("USAGE:")
+		fmt.Println("  local-sensor <minutes> <output file path>")
+		log.Fatalf("Incorrect number of arguments, expected 2 but found: %d", len(os.Args)-1)
+	}
+
+	i, err := strconv.Atoi(os.Args[1])
+	if err != nil {
+		log.Fatalf("First parameter must be a valid integer")
+	}
+
+	return time.Duration(i) * time.Minute, path.Clean(os.Args[2])
+}
+
+func registerHostKillSignals(startTime time.Time, fakeCentral *centralDebug.FakeService, outfile string) {
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-c
+		endTime := time.Now()
+		allMessages := fakeCentral.GetAllMessages()
+		dumpMessages(allMessages, startTime, endTime, outfile)
+		os.Exit(0)
+	}()
+}
```

Comment:
How about something context-based? (if needed, one could add a timeout to the context as well)

```suggestion
func registerHostKillSignals(startTime time.Time, fakeCentral *centralDebug.FakeService, outfile string) {
	ctx, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
	defer stop()
	for {
		select {
		case <-ctx.Done():
			endTime := time.Now()
			allMessages := fakeCentral.GetAllMessages()
			dumpMessages(allMessages, startTime, endTime, outfile)
			os.Exit(0)
		}
	}
}
```

Reply:
Sorry, the suggestion requires to start `registerHostKillSignals` in a goroutine - forgot to add this.

Reply:
I don't understand why looping on a non-blocking `select` would be better. The suggestion you propose keeps executing the block under `default` while the os signal is not sent.

I can't see why we wouldn't want to block the goroutine here while this context is not done, like:
```go
func registerHostKillSignals(startTime time.Time, fakeCentral *centralDebug.FakeService, outfile string) {
        ctx, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
        defer stop()
        <- ctx.Done()
        endTime := time.Now()
        allMessages := fakeCentral.GetAllMessages()
        dumpMessages(allMessages, startTime, endTime, outfile)
        os.Exit(0)

}

```

Reply:
Yeah, after trying that locally, I think you are right - the benefit is minuscule and to leverage it, we would need to replace `context.Background()` with something that expires after a timeout.